numpyro/util.py

```diff
@@ -115,31 +115,26 @@ def scan(f, a, bs):
         return lax.scan(f, a, bs)
 
 
-def tscan(f, a, bs, fields=(0,)):
+def tscan(f, a, bs, transform):
+    if not transform:
+        def transform(x): return x
     if _DISABLE_CONTROL_FLOW_PRIM:
         length = tree_flatten(bs)[0][0].shape[0]
         for i in range(length):
             b = tree_map(lambda x: x[i], bs)
             a = f(a, b)
-            a_out = tree_map(lambda x: x[None, ...], a)
-
-            # the following three lines are necessary for tscan;
-            # it might be useful in the case you want to use `transform` instead of `fields`
-            a_flat, a_tree = tree_flatten(a_out)
-            a_selected = [field if i in fields else None for i, field in enumerate(a_out)]
-            a_out = tree_unflatten(a_tree, a_selected)
-
+            a_out = transform(tree_map(lambda x: x[None, ...], a))
```

Comment:
Neat! I am surprised that it is so simple. Could you add a fix for me about the case `x` is scalar here? That is to change `lambda x:...` to `lambda x: np.expand_dims(x, axis=0)`.

Reply:
More concretely, it is to resolve: `np.array(1., dtype=np.float32)[None, ...]` works but ``np.array(1.)[None, ...]` does not work as expected.

Reply:
I saw that in `lax.scan` too. I think we should only expand for the scalar case, right?

Reply:
`lax.scan` should be fine because x will be a DeviceArray there (see my above comment)

Reply:
I meant, something like this `lax.scan(f, np.array(0.), np.array([0., 1., 2.]))` throws an exception for the same reason, but it won't if you add a `dtype=float32` to the first arg. 

I think the issue is that `numpy.array(0.)` is actually a scalar array but `jax.numpy.array(0.)` is a python float without the `dtype` arg.

Reply:
Yes, you are right. But we might not face it when using `scan` with HMC. I remember that I get this error while using `tscan` to debug for covtype example with prims disabled, adapt_step_size=False. So `step_size` is float during the tscan. Somehow using `lax.scan` is fine because `step_size` will be converted to a DeviceArray during the way.

Reply:
I think this is a regression in JAX - https://github.com/google/jax/issues/121, maybe? In any case, I'll put in a work-around for this.